scripts: get_maintainer: Add sub-area support

decsny opened the pull request:
Add support to specify "parents" key on an area, to link it as a child of another area. Change path2areas to recursively add all parents of areas to a path. Add a command to view the hierachy of areas easily.

This is based on my understanding of the description @aescolar gave me about "sub-areas" that are mentioned at the top of the MAINTAINER.yml but seem to not actually have any support.

Essentially, each area will now be appended with the labels of its parent areas, and each file path will be associated with not only the direct area matches but all parents and grandparentc etc of the area matches.

This could also theoretically allow for more intelligent and complex assignment algorithms and processes in the future but that discussion is out of the scope of this PR.

Also, the reason I chose to make a "parents" key on the child areas instead of a "children" key on the parent areas, is because then it is easier to specify multiple-inheritance of areas. For example, a USB driver could be both a child of a vendor area and the USB generic area, and therefore only need to edit one place instead of two.

TODO: Still need to figure out the interaction with files-exclude

Commit:
scripts: get_maintainer: Add sub-area support

Add support to specify "parents" key on an area, to link it as a child
of another area. Change path2areas to recursively add all parents of
areas to a path. Add a command to view the hierachy of areas easily.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

henrikbrixandersen commented:
@decsny Perhaps add a subarea (or convert [an] existing area[s]) to subareas in `MAINTAINERS.yml` in this PR as well to demonstrate its use?

Commit:
MAINTAINERS: Use sub-areas for disk area

Use new sub-area feature for Disk Area

get_maintainers.py hierarchy Disk :

Disk
├── Drivers: Disk
└── SD
    ├── Drivers: SDHC
    └── MMC

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
MAINTAINERS: Rework NXP area with sub-areas

Rework NXP area with new sub-areas

NXP Drivers
├── NXP BLE Drivers
├── NXP IEEE802154
├── NXP MCUX USB Driver
└── NXP Wifi Drivers

NXP Platforms (MCU)
├── NXP Kinetis Platforms
├── NXP LPC Platforms
├── NXP MCU Documentation
├── NXP Platforms (Robotics Products)
├── NXP Platforms (Xtensa)
└── NXP Wireless MCUs

NXP Platforms (MPU)
└── NXP Platforms (Xtensa)

NXP Platforms (S32)
└── NXP Platforms (Robotics Products)

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
I updated the PR with example of Disk areas, and NXP areas using the parents key. I put the output of the new hierarchy command in the commit message.
